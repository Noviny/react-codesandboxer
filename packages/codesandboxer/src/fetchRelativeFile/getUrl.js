// @flow
import type { GitInfo } from '../types';

const raw = {
  github: (path, { account, repository, branch = 'master' }) =>
    `https://raw.githubusercontent.com/${account}/${repository}/${branch}/${path}`,
  bitbucket: (path, { account, repository, branch = 'master' }) =>
    `https://api.bitbucket.org/1.0/repositories/${account}/${repository}/raw/${branch}/${path}`,
};

export default function getUrl(
  path: string,
  { host, ...urlConfig }: GitInfo,
) {
  let getRaw = raw[host];
  if (typeof getRaw !== 'function')
    throw new Error(`Could not parse files from ${host}`);

  let url = getRaw(path, urlConfig);
  let fileType = '';
  let fileMatch = path.match(/.+(\..+)$/);
  if (!fileMatch) {
    return { fileType: '.js', url: `${url}.js` };
  } else {
    return { fileType: fileMatch[1], url };
  }
}
