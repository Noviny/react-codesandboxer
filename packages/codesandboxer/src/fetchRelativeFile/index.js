// @flow
import parseFile from '../parseFile';
import replaceImports from '../replaceImports';
import absolutesToRelative from '../utils/absolutesToRelative';
import getUrl from './getUrl';
import 'isomorphic-unfetch';
import type {
  Package,
  GitInfo,
  ParsedFile,
  Config,
} from '../types';

/*
This is modified from the canvas answer here:
https://stackoverflow.com/questions/6150289/how-to-convert-image-into-base64-string-using-javascript
*/
function fetchImage(url, path): Promise<ParsedFile> {
  return new Promise((resolve) => {
    var img = new Image();
    img.crossOrigin = 'Anonymous';
    img.src = url;

    img.onload = function() {
      var canvas: HTMLCanvasElement = document.createElement('canvas');
      var ctx = canvas.getContext('2d');
      var dataURL;
      canvas.height = this.naturalHeight;
      canvas.width = this.naturalWidth;
      ctx.drawImage(this, 0, 0);
      dataURL = canvas.toDataURL();
      resolve(dataURL);
    };
  }).then(file => ({ file, deps: {}, internalImports: [], path }));
}

const fetchJS = (url, path, pkg, importReplacements): Promise<ParsedFile> => {
  return (
    fetch(url)
      .then(res => {
        if (res.status === 404) {
          throw new Error(`file not found at: ${url}`);
        }
        return res.text();
      })
      .then(content =>
        replaceImports(
          content,
          importReplacements.map(m => [absolutesToRelative(path, m[0]), m[1]]),
        ),
      )
      // this is not correct
      .then(content => parseFile(content, pkg))
      .then(file => ({ ...file, path }))
  );
};

const fetchJSON = (url, path): Promise<ParsedFile> => {
  return fetch(url)
    .then(res => {
      if (res.status === 404) {
        throw new Error(`file not found at: ${url}`);
      }
      return res.text();
    })
    .then(file => ({ file, deps: {}, internalImports: [], path }));
};

/*
resolution order:
A.js
A.json
A.jsx (if allowed)
A/index.js
A/index.json
A/index.jsx (if allowed)
*/

const catchBlock = (e, url, path, pkg, importReplacements, extension) => {
  if (e.message.includes('file not found at:')) {
    let newPath = `${path}${extension}`;
    let newUrl = url.replace(/.js$/, extension);
    if (extension === '.json') {
      return fetchJSON(newUrl, newPath);
    }
    return fetchJS(newUrl, newPath, pkg, importReplacements);
  } else {
    throw e;
  }
};

// Imports that are not named may be .js, .json, or /index.js. Node resolves them
// in that order.
let fetchProbablyJS = (url, path, pkg, importReplacements, config) => {
  // This first fetch handles .js
  return fetchJS(url, path, pkg, importReplacements)
    .catch(e => catchBlock(e, url, path, pkg, importReplacements, '.json'))
    .catch(e => {
      if (config.allowJSX)
        {return catchBlock(e, url, path, pkg, importReplacements, '.jsx');}
      else throw e;
    })
    .catch(e => catchBlock(e, url, path, pkg, importReplacements, '/index.js'))
    .catch(e =>
      catchBlock(e, url, path, pkg, importReplacements, '/index.json'),
    )
    .catch(e => {
      if (config.allowJSX)
        {return catchBlock(e, url, path, pkg, importReplacements, '/index.jsx');}
      else throw e;
    });
};

let fetchFileContents = (
  url,
  path,
  { fileType, pkg, importReplacements },
  config,
): Promise<ParsedFile> => {
  switch (fileType) {
    case '.png':
    case '.jpeg':
    case '.jpg':
    case '.gif':
    case '.bmp':
    case '.tiff':
      return fetchImage(url, path);
    case '.js':
      return fetchProbablyJS(url, path, pkg, importReplacements, config);
    case '.jsx':
      return fetchJS(url, path, pkg, importReplacements);
    case '.json':
      return fetchJSON(url, path);
    default:
      throw new Error(`unparseable filetype: ${fileType} for file ${path}`);
  }
};

type HandleFileFetch = Promise<ParsedFile>;

export default async function fetchRelativeFile(
  path: string,
  pkg: Package,
  importReplacements: Array<[string, string]>,
  gitInfo: GitInfo,
  config?: Config,
): HandleFileFetch {
  config = config || {};
  // The new path is the file name we will provide to codesandbox
  // Get the url from the gitInfo. For JS files, we will need to add the filetype
  // This method needs to determine the filetype, so we return it.
  let { url, fileType } = getUrl(path, gitInfo);

  let file = await fetchFileContents(
    url,
    path,
    {
      fileType,
      pkg,
      importReplacements,
    },
    config,
  );
  return file;
}
