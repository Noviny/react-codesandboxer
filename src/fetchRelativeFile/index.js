// @flow
import resolvePath from '../utils/resolvePath';
import parseFile from '../parseFile';
import replaceImports from '../replaceImports';
import absolutesToRelative from '../utils/absolutesToRelative';
import getUrl from './getUrl';
import type {
  Package,
  FetchConfig,
  ParsedFile,
  Dependencies,
  Import,
} from '../types';

/*
This is modified from the canvas answer here:
https://stackoverflow.com/questions/6150289/how-to-convert-image-into-base64-string-using-javascript
*/
const supportedFiles = ['.js', '.json'];

const supportedImageFormats = [
  '.png',
  '.jpeg',
  '.jpg',
  '.gif',
  '.bmp',
  '.tiff',
];

function fetchImage(url, path): Promise<ParsedFile> {
  return new Promise((resolve, reject) => {
    var img = new Image();
    img.crossOrigin = 'Anonymous';
    img.src = url;

    img.onload = function() {
      var canvas: HTMLCanvasElement = document.createElement('canvas');
      var ctx = canvas.getContext('2d');
      var dataURL;
      canvas.height = this.naturalHeight;
      canvas.width = this.naturalWidth;
      ctx.drawImage(this, 0, 0);
      dataURL = canvas.toDataURL();
      resolve(dataURL);
    };
  }).then(file => ({ file, deps: {}, internalImports: [] }));
}

const fetchJS = (url, path, pkg, importReplacements): Promise<ParsedFile> => {
  importReplacements.map(m => [absolutesToRelative(path, m[0]), m[1]]);
  return fetch(url)
    .then(a => a.text())
    .then(content =>
      replaceImports(
        content,
        importReplacements.map(m => [absolutesToRelative(path, m[0]), m[1]]),
      ),
    )
    .then(content => parseFile(content, pkg));
};

const fetchJSON = (url, path): Promise<ParsedFile> => {
  return fetch(url)
    .then(res => res.text())
    .then(file => ({ file, deps: {}, internalImports: [] }));
};

let fetchFileContents = (
  url,
  path,
  { isImage, fileType, pkg, importReplacements },
): Promise<ParsedFile> => {
  switch (fileType) {
    case '.png':
    case '.jpeg':
    case '.jpg':
    case '.gif':
    case '.bmp':
    case '.tiff':
      return fetchImage(url, path);
    case '.js':
      return fetchJS(url, path, pkg, importReplacements);
    case '.json':
      return fetchJSON(url, path);
    default:
      throw new Error(`unparseable filetype: ${fileType} for file ${path}`);
  }
};

type HandleFileFetch = Promise<{
  file: string,
  deps: Dependencies,
  internalImports: Array<string>,
}>;

export default async function fetchRelativeFile(
  path: string,
  pkg: Package,
  importReplacements: Array<[string, string]>,
  config: FetchConfig,
): HandleFileFetch {
  // The new path is the file name we will provide to codesandbox
  // Get the url from the config. For JS files, we will need to add the filetype
  // This method needs to determine the filetype, so we return it.
  let { url, fileType } = getUrl(path, config);

  let file = await fetchFileContents(url, path, {
    fileType,
    pkg,
    importReplacements,
  });
  return { ...file };
}
