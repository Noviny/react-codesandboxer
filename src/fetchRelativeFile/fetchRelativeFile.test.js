import cases from 'jest-in-case';
import fetchRelativeFile from './';
import getUrl from './getUrl';
import isomorphic from 'isomorphic-fetch';
const BBConfig = {
  account: 'noviny',
  repository: 'react-codesandboxer',
  branch: 'ca41dfb340eed67c7f755e0b03939652d1f42cc7',
  host: 'github',
};

describe('fetchRelativeFile()', () => {
  it('should fetch a .js file in a subdirectory', async () => {
    return fetchRelativeFile(
      'src/CodeSandboxer/index.js',
      {},
      [],
      BBConfig,
    ).then(({ file, deps, internalImports }) => {
      expect(deps).toMatchObject({ react: 'latest' });
      expect(internalImports).toEqual(
        expect.arrayContaining(['../types', '../fetchFiles']),
      );
      expect(file).toMatchSnapshot();
    });
  });
  it('should fetch a .js file at the root');
  it('should fetch a .png file in a subdirectory');
  it('should fetch a .png file at the root');
  it('should fetch a .json file in a subdirectory');
  it('should fetch a .json file at the root', async () => {
    return fetchRelativeFile('package.json', {}, [], BBConfig).then(
      ({ file, deps, internalImports }) => {
        expect(deps).toMatchObject({});
        expect(internalImports).toEqual(expect.arrayContaining([]));
        expect(file).toMatchSnapshot();
      },
    );
  });
});

cases(
  'getUrl()',
  ({ path, expectedType }) => {
    const { fileType, url } = getUrl(path, BBConfig);
    expect(fileType).toBe(expectedType);
  },
  [
    { name: 'js file', path: 'something.js', expectedType: '.js' },
    { name: 'png file', path: 'something.png', expectedType: '.png' },
    {
      name: 'png file extended',
      path: 'abc/something.png',
      expectedType: '.png',
    },
    {
      name: 'png file extended',
      path: 'abc/something.json',
      expectedType: '.json',
    },
    {
      name: 'root package.json fetch',
      path: 'package.json',
      expectedType: '.json',
    },
  ],
);
