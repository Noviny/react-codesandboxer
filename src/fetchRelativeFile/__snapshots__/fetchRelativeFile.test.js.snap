// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchRelativeFile() should fetch a .js file in a subdirectory 1`] = `
"// @flow
import React, { Component, type Node } from 'react';
import type { Package, Files, FetchConfig } from '../types';
import fetchFiles from '../fetchFiles';

const codesandboxURL = 'https://codesandbox.io/api/v1/sandboxes/define';

type State = {
  parameters: string,
};

type Props = {
  /* The absolute path to the example within the git file structure */
  examplePath: string,
  /* This is all the information we need to fetch information from github or bitbucket */
  gitInfo: FetchConfig,
  example?: string | Promise<string>,
  /* Either take in a package.json object, or a string as the path of the package.json */
  pkgJSON?: Package | string | Promise<Package | string>,
  /* paths in the example that we do not want to be pulled from their relativeLocation */
  importReplacements: Array<[string, string]>,
  /* Dependencies we always include. Most likely react and react-dom */
  dependencies?: { [string]: string },
  /* Do not actually deploy to codesanbox. Used to for testing alongside afterDeploy */
  skipDeploy?: boolean,
  ignoreInternalImports?: boolean,
  /* function that can be called once the deploy has occurred, useful if you want to give feedback or test how CSB is working */
  afterDeploy?: (
    { parameters: string, files: Files } | { error: any },
  ) => mixed,
  /* Pass in files separately to fetching them. Useful to go alongisde specific replacements in importReplacements */
  providedFiles?: Files,
  children?: Node,
};

export default class CodeSandboxDeployer extends Component<Props, State> {
  form: HTMLFormElement | null;

  state = { parameters: '' };
  static defaultProps = {
    children: <button type=\\"submit\\">Deploy to CodeSandbox</button>,
    dependencies: {},
    providedFiles: {},
    importReplacements: [],
  };

  deployToCSB = (e: MouseEvent) => {
    const {
      examplePath,
      pkgJSON,
      gitInfo,
      dependencies,
      skipDeploy,
      afterDeploy,
      importReplacements,
      providedFiles,
      example,
    } = this.props;
    e.preventDefault();

    fetchFiles(
      examplePath,
      pkgJSON,
      gitInfo,
      importReplacements,
      dependencies,
      providedFiles,
      example,
    )
      .then(({ parameters, files }) => {
        this.setState({ parameters }, () => {
          if (!skipDeploy && this.form) this.form.submit();
          if (afterDeploy) afterDeploy({ parameters, files });
        });
      })
      .catch(error => {
        if (afterDeploy) afterDeploy({ error });
      });
  };

  render() {
    return (
      <form
        style={{ display: 'inline-block' }}
        onSubmit={this.deployToCSB}
        action=\\"https://codesandbox.io/api/v1/sandboxes/define\\"
        method=\\"POST\\"
        target=\\"_blank\\"
        ref={r => {
          this.form = r;
        }}
      >
        <input type=\\"hidden\\" name=\\"parameters\\" value={this.state.parameters} />
        {this.props.children}
      </form>
    );
  }
}
"
`;

exports[`fetchRelativeFile() should fetch a .json file at the root 1`] = `
Object {
  "author": "Ben Conolly",
  "dependencies": Object {
    "codesandbox": "^1.1.14",
  },
  "description": "A simple react component to help easily deploy an example to codesandbox",
  "devDependencies": Object {
    "babel-cli": "^6.26.0",
    "babel-plugin-syntax-dynamic-import": "^6.18.0",
    "babel-plugin-transform-class-properties": "^6.24.1",
    "babel-plugin-transform-object-rest-spread": "^6.26.0",
    "babel-plugin-transform-runtime": "^6.23.0",
    "babel-preset-env": "^1.6.1",
    "babel-preset-flow": "^6.23.0",
    "babel-preset-react": "^6.24.1",
    "flow-bin": "^0.64.0",
    "jest": "^22.1.4",
    "jest-in-case": "^1.0.2",
    "prettier": "^1.10.2",
    "react": "16.2.0",
  },
  "files": Array [
    "dist",
    "README.md",
  ],
  "keywords": Array [
    "codesandbox",
    "react",
    "react-component",
  ],
  "license": "MIT",
  "main": "dist/index.js",
  "name": "react-codesandboxer",
  "peerDependencies": Object {
    "react": "^0.14.9 || ^15.3.0 || ^16.0.0-rc || ^16.0",
    "react-dom": "^0.14.9 || ^15.3.0 || ^16.0.0-rc || ^16.0",
  },
  "repository": Object {
    "type": "git",
    "url": "https://github.com/Noviny/react-codesandboxer.git",
  },
  "scripts": Object {
    "build": "babel src -d dist --ignore __tests__/",
    "dev": "yarn build --watch",
    "prepublishOnly": "yarn test && yarn flow && yarn build",
    "test": "yarn jest",
  },
  "version": "0.2.1",
}
`;
